diff --git a/src/libstore/build/sandbox-defaults.sb b/src/libstore/build/sandbox-defaults.sb
index 77f013aea..0283695b4 100644
--- a/src/libstore/build/sandbox-defaults.sb
+++ b/src/libstore/build/sandbox-defaults.sb
@@ -107,4 +107,7 @@ R""(
        (subpath "/System/Library/LaunchDaemons/com.apple.oahd.plist")
        (subpath "/Library/Apple/System/Library/LaunchDaemons/com.apple.oahd.plist"))

+; Allow the system OpenSSL to read its config on 14.2+.
+(allow file-read* (literal "/private/etc/ssl/openssl.cnf"))
+
 )""
